src/gentropy/common/harmonise.py

```diff
@@ -0,0 +1,180 @@
+    chromosome: str,
+    colname_position: str,
+    colname_allele0: str,
+    colname_allele1: str,
+    colname_a1freq: str,
+    colname_info: str,
+    colname_beta: str,
+    colname_se: str,
+    colname_mlog10p: str,
+    colname_n: str,
+) -> DataFrame:
+    """Ingest and harmonise the summary stats.
+
+    1. Rename chromosome 23 to X.
+    2. Filter out low INFO rows.
+    3. Filter out low frequency rows.
+    4. Assign variant types.
+    5. Create variant ID for joining the variant annotation dataset.
+    6. Join with the Variant Annotation dataset.
+    7. Drop bad quality variants.
+
+    Args:
+        spark (SparkSession): Spark session object.
+        raw_summary_stats_path (str): Input raw summary stats path.
+        tmp_variant_annotation_path (str): Input variant annotation dataset path.
+        chromosome (str): Which chromosome to process.
+        colname_position (str): Column name for position.
+        colname_allele0 (str): Column name for allele0.
+        colname_allele1 (str): Column name for allele1.
+        colname_a1freq (str): Column name for allele1 frequency (optional).
+        colname_info (str): Column name for INFO, reflecting variant quality (optional).
+        colname_beta (str): Column name for beta.
+        colname_se (str): Column name for beta standard error.
+        colname_mlog10p (str): Column name for -log10(p).
+        colname_n (str): Column name for the number of samples.
+
+    Returns:
+        DataFrame: A harmonised summary stats dataframe.
+    """
+    # Read the precomputed variant annotation dataset.
+    va_df = (
+        spark
+        .read
+        .parquet(tmp_variant_annotation_path)
+        .filter(f.col("vaChromosome") == ("X" if chromosome == "23" else chromosome))
+        .persist()
+    )
+
+    # Read and process the summary stats dataset.
+    df = (
+        spark
+        .read
+        .parquet(raw_summary_stats_path)
+        .filter(f.col("chromosome") == chromosome)
+        # Harmonise, 1: Rename chromosome 23 to X.
+        .withColumn(
+            "chromosome",
+            f.when(
+                f.col("chromosome") == "23", "X"
+            ).otherwise(f.col("chromosome"))
+        )
+    )
+    if colname_info:
+        # Harmonise, 2: Filter out low INFO rows.
+        df = df.filter(f.col(colname_info) >= 0.8)
+    if colname_a1freq:
+        # Harmonise, 3: Filter out low frequency rows.
+        df = (
+            df
+            .withColumn(
+                "MAF",
+                f.when(f.col(colname_a1freq) < 0.5, f.col(colname_a1freq))
+                .otherwise(1 - f.col(colname_a1freq))
+            )
+            .filter(f.col("MAF") >= 0.0001)
+            .drop("MAF")
+        )
+    df = (
+        df
+        # Harmonise, 4: Assign variant types.
+        .withColumn(
+            "variant_type",
+            f.when(
+                (f.length(colname_allele0) == 1) & (f.length(colname_allele1) == 1),
```

Comment:
Please add short memo on the variant_type, I feel I have asked about it last time I reviewed the code, but can not remember again what was the rationale.

Reply:
Agreed, that'a s good point. Now added a comment outlining three possible variant types